Another take on projectionist config

kelvinst opened the pull request:
This is just to incremento your PR https://github.com/elixir-editors/vim-elixir/pull/451 with the stuff @hauleth commented + some dispatch config.

Also, I changed it so it to change the `heuristics` option instead of adding the projections on the existent ones, because the way you had it before was not possible to use `:Pcd` to change directories to the closest `mix.exs` file.

Thanks!

mluts committed:
Add basic projectionist support

schnittchen committed:
Folding tests (#453)

* basic folding test

* move matcher over, extract buffer method, example dsl

* support folding on lower lines, give failure message

* add a few examples

jbodah committed:
replace all indent tests with i helper (#455)

feliperenan committed:
Fix syntax highlighting with "~S" sigil (#457)

Fixes https://github.com/elixir-editors/vim-elixir/issues/438

jparise committed:
Add .xrl to the suffixes list (#461)

This is the suffix used by leex, Erlang's lexer generator.

veita committed:
Fix issue-459 (#460)

Issue 459: manual_install.sh creates wrong directory hierarchy.

Commit:
add failing spec for implicit rescues

Commit:
Revert "add failing spec for implicit rescues"

This reverts commit da5bc2c2693febf5c1fa34aba08169ecb7aee75f.

jbodah committed:
try else indent (#462)

* add failing try/else spec

* add indent rules for try/else

Commit:
remove pending flag from test

Commit:
remove pending guard syntax tests

jbodah committed:
fix if/else folding bug (#468)

* rework folding spec helper

* remove elixirElseBlock

* remove unsupported else fold spec

Commit:
speed up fold tests

Commit:
speed up vim start

Commit:
micro optimizations for tests

jparise committed:
Use 'set filetype' to set filetype once again (#467)

The s:setf() helper function was originally introduced to avoid setting
the filetype multiple times (375e9e1). I believe this covered the case
where a file had an .ex or .exs extension and also include "elixir" in
the shebang line (such that s:DetectElixir() would match and call 'set
filetype' redundantly).

c56bd82 removed the existing &filetype check in s:setf(), relying on the
&filetype global and ftdetect framework to avoid redundant updates.

I believe we can simplify this even further by using just 'set filetype'
in the file extension-based autocommands and including a guard against
(re)setting filetype in s:DetectElixir().

Also, use explicit case-insensitive comparison operators. This isn't a
good place to make assumptions about 'ignorecase'.

lukebakken committed:
Add vim 8 package manager instructions (#449)

See the following links for details:

http://vimhelp.appspot.com/repeat.txt.html#packages
https://shapeshed.com/vim-packages/

Commit:
Syntax: Add doc delimiters (#470)

Why
---

I want to be able to differentiate between ordinary delimiters and
those used for docs

How
---

* Add doc syntax specs
* Add **elixirDocSigilDelimiter** group
* Add **elixirDocStringDelimiter** group
* Link **elixirDocStringDelimiter** and **elixirDocStringDelimiter** to
  **elixirStringDelimiter** and **elixirSigilDelimiter** to preserve
  existing behavior

jparise committed:
Remove redundant 'formatoptions' setting (#472)

'formatoptions' is already set (to these same values) on line 35.

Also, add 'fo<' to the 'b:undo_ftplugin' list.

jparise committed:
Use 'setfiletype' to set local &filetype (#473)

'setfiletype' is nice in these cases because it prevents &filetype from
being set redundantly.

Note that the Vim runtime ships with a filetype match for *.ex (for
Euphoria), but our plugin-based filetype detection still gets the last
say.

Also, use 'setlocal' when setting &filetype in the shebang detection
path. We don't want to use 'setfiletype' here because it's possible we
need to override an existing syntax assignment, and we already guard
against setting &filetype to 'elixir' twice.

jbodah committed:
Revert "Use 'setfiletype' to set local &filetype (#473)" (#476)

This reverts commit 653cc4d24183558833ee9942cf14731b740d5a55.

jbodah committed:
split up write/redraw in test suite to improve reliability

Commit:
Syntax: Add private declarations (#471)

Why
---

I want to be able to differentiate between public and private
declarations

How
---

* Add **elixirPrivateFunctionDeclaration** group
* Add **elixirPrivateRecordDeclaration** group
* Add **elixirPrivateMacroDeclaration** group
* Link **elixirPrivateFunctionDeclaration** to
  **elixirFunctionDeclaration** to preserve behavior
* Link **elixirPrivateRecordDeclaration** to **elixirRecordDeclaration**
  to preserve behavior
* Link **elixirPrivateMacroDeclaration** to **elixirMacroDeclaration**
  to preserve behavior
* Add **elixirPrivateFunctionDeclaration**,
  **elixirPrivateRecordDeclaration**, and
  **elixirPrivateMacroDeclaration** to **elixirDeclaration** cluster

dreamtigers committed:
Fix wrong highlight of paren after `defmodule` (#478)

Added to `elixirModuleDeclaration` the exact same regex used in
`elixirFunctionDeclaration`.

navargas committed:
Update README.md vim 8 install instructions (#480)

Vim 8 reads plugins from `~/.vim/pack/my-packages/start/pluginName`

hauleth committed:
vim-dadbod integration for applications using Ecto (#481)

* feat: vim-dadbod integration for applications using Ecto

[vim-dadbod][1] is useful library for interacting with databases, this
PR adds support for interacting with applications that are using Ecto
with `Ecto.Adapter.Postgres` and `Ecto.Adapter.MySQL`, thanks to that
user can run:

    :DB ecto:MyApp.Repo SELECT * FROM foo

And it will open new preview window containing result of the query or
even one can use

    :DB ecto:MyApp.Repo

To open interactive shell for DB assigned to given repository.

[1]: https://github.com/tpope/vim-dadbod

* docs: include informations about integration in README

cadebward committed:
add support for live view templates (#483)

liskin committed:
syntax: Allow escaping of doc heredocs (#489)

* syntax: Align matches in elixirDocString

Broken since cec4aad4a587650ab9fcd50f9d2a17e91bf648da.

* test: retab spec/syntax/doc_spec.rb

* syntax: Allow escaping of doc heredocs

Use the same end pattern as normal heredocs.

Fixes #469

jbodah committed:
Add Phoenix LiveEEx Template (.leex) file type support (2) (#490)

* Add Phoenix LiveEEx Template (.leex) file type support.
Add LiveView sigil (~L"""LiveView""") support in .ex.
Update README.md

* update leex indentation and syntax specs
removed duplicated live view sigil spec

* fix leex test bugs

liskin committed:
ftdetect: Fix detection when reopening

When doing :e in an already opened elixir script lacking the .exs
filename suffix, we would skip the check and end up with filetype=conf
or something like that.

Fix this by forcing the check when `!did_filetype()` even if `&filetype ==# 'elixir'`.

Commit:
add docker container file with isolated vim instance

jbodah committed:
Update README.md

liskin committed:
syntax: Fix uppercase Atom: in keyword lists

This was broken e.g. for
https://github.com/elixir-lang/elixir/blob/master/lib/elixir/docs.exs.

liskin committed:
syntax: Highlight dots in aliases as aliases

It's confusing that "defmodule Foo.Bar" was highlighted differently than
"Foo.Bar.fun". Unify the two.

liskin committed:
syntax: Prevent syntax/html.vim from overriding our syn sync

When including other syntax types it's necessary to declare that we're
the main syntax, otherwise they will apply their own syn sync settings.

Here, by including html.vim in 7e00da6033b7847c6bb71df18f852342946eab42,
we end up with syn sync minlines=10 maxlines=100 and we sync using
html.vim syncing rules, which isn't at all what we want. Try opening
large_file.ex in the middle, syntax will be incorrect.

liskin committed:
syntax: Prevent syntax/javascript.vim from overriding our syn sync

Unfortunately the main_syntax trick doesn't prevent javascript.vim from
doing syn sync fromstart. This is most likely a bug in javascript.vim
and I'll submit a pull request to vim to fix this. Luckily setting
main_syntax does prevent javascript.vim from introducing its own syncing
rules, it just overrides minlines and maxlines, so we can work around it
by moving the syn sync minlines to the end.

liskin committed:
syntax: Fix elixir_use_markdown_for_docs

Most (if not all) syntax files unconditionally set b:current_syntax at
the end, so we need to unlet it, otherwise the second inclusion is
skipped.

Commit:
add nvim to docker-compose

dbernheisel committed:
Set mix.lock as Elixir

ujihisa committed:
Introduce initial doc

* https://github.com/elixir-editors/vim-elixir/issues/496

ujihisa committed:
Do not change &formatprg globally

Since `&formatprg` is "global or local to buffer", `:set` does not only change &formatprg of the current buffer, but also any other buffers. Prefer `:setlocal` instead.

```
:h formatprg
						*'formatprg'* *'fp'*
'formatprg' 'fp'	string (default "")
			global or local to buffer |global-local|
	The name of an external program that will be used to format the lines
	selected with the |gq| operator.  The program must take the input on
	stdin and produce the output on stdout.  The Unix program "fmt" is
	such a program.
	If the 'formatexpr' option is not empty it will be used instead.
	Otherwise, if 'formatprg' option is an empty string, the internal
	format function will be used |C-indenting|.
	Environment variables are expanded |:set_env|.  See |option-backslash|
	about including spaces and backslashes.
	This option cannot be set from a |modeline| or in the |sandbox|, for
	security reasons.
```

kassio committed:
Ensure right string matching case for elixirExUnit* patterns

kelvinst committed:
Using mix compile, not mix as makeprg for mix compiler

kelvinst committed:
Improving the error format for mix compile

kelvinst committed:
Another take on projectionist config

hauleth commented on ftplugin/elixir.vim:
```diff
@@ -52,18 +52,52 @@ silent! setlocal formatoptions-=t formatoptions+=croqlj
+        \       }
+        \     }
+        \   }
+        \ })
```
I would use 3rd argument with value `keep` to not overwrite user-defined projections.

kelvinst replied:
Yep, doing it

hauleth commented on ftplugin/elixir.vim:
```diff
@@ -52,18 +52,52 @@ silent! setlocal formatoptions-=t formatoptions+=croqlj
+        \         'defmodule {camelcase|capitalize|dot}Test do',
+        \         '  use ExUnit.Case',
+        \         '',
+        \         '  alias {camelcase|capitalize|dot}, as: Subject',
```
I do not think that this line is common practice. Even I am questioning it currently.

kelvinst replied:
Ture, I will remove the alias and the reference to it.

kelvinst committed:
Reading column number on credo errors

Commit:
Fix highlight for atoms containing @ character

jbodah committed:
Revert "Fix highlight for atoms containing @ character"

This reverts commit eb06df6d02cb1deb07af66d33a4aedd6e8769e0b.

jbodah committed:
Fix doc sigil with paren

Fixes #477

jbodah committed:
don't change indent of doc blocks, strings, and comments

Fixes #505
Fixes #346

hauleth committed:
Syntax: add __STACKTRACE__ (#514)

oleks committed:
Use /usr/bin/env to find bash

On some systems, bash does not lie under /bin/. For instance, NixOS.

pksunkara committed:
Add ctags to gitignore

I submodule this repo to use it with pathogen and it keeps breaking when I update because of untracked content in the submodule. This should fix the issue

sgiath committed:
Add raise as keyword

Closes #519

Signed-off-by: Sgiath <FilipVavera@sgiath.dev>

sgiath committed:
Move test to one it

Signed-off-by: Sgiath <FilipVavera@sgiath.dev>

jbodah committed:
add elixirFunctionCall

Implementation from https://github.com/elixir-editors/vim-elixir/pull/506

cc: Thanks @w-sanches

aoswalt committed:
Add tilde operators syntax matches

jbodah committed:
light brush up of README

jbodah committed:
add feature wishlist to README

jbodah committed:
tweak maintenance help wording

jbodah committed:
add configuration tips to README

szTheory committed:
README.md typo fix

szTheory committed:
README.md update git clone URLs

kelvinst commented:
Closing it as I just don't use my projectionist config like that anymore.
